### PR TITLE
Initial patch with Jedi completion (no function header description).

### PR DESCRIPTION
**Pull request by liukelly**

### CODE REVIEW

**Comment by Carreau** [comment]
Thanks , this is a first part toward #8606 , @liukelly is working with me from BIDS.

File: IPython/core/completer.py [hunk 1]
```diff
@@ -765,9 +775,59 @@ def magic_matches(self, text):
...
+            sys.path.pop(0)
+
+        with open('tmp.txt', 'w') as f:
+            print('before:', before, file=f)
+            for c in completions:
+                print('', file=f)
+                print('full_name:', c.full_name, file=f)
+                print('name:', c.name, file=f)
+                print('name_with_symbols:', c.name_with_symbols, file=f)
+                print('description:', c.description, file=f)
```
Comment (Carreau) [inline IPython/core/completer.py]: That's for debug, we need to figure out why `description:` is always `module:`

File: IPython/core/completer.py [hunk 2]
```diff
@@ -76,6 +76,13 @@
 from IPython.utils.py3compat import builtin_mod, string_types, PY3
 from traitlets import CBool, Enum
 
+try:
+    import jedi
+    import jedi.api.helpers
+    import jedi.parser.user_context
+except ImportError:
+    pass
```
Comment (Carreau) [inline IPython/core/completer.py]: From in person discussion: handle the case where Jedi is not importable, by skipping the jedi codepath.

File: IPython/core/completer.py [hunk 3]
```diff
@@ -765,9 +775,59 @@ def magic_matches(self, text):
...
+        if self.namespace is None:
+            import __main__
+            namespaces.append(__main__.__dict__)
+        else:
+            namespaces.append(self.namespace)
+        if self.global_namespace is not None:
+            namespaces.append(self.global_namespace)
+
+        # From Jedi docs
+        sys.path.insert(0, os.getcwd())
```
Comment (Carreau) [inline IPython/core/completer.py]: we probably want to avoid manipulating path here.

**Comment by Carreau** [comment]
So the description think seem to be a Jedi Bug, I opened an issue:

https://github.com/davidhalter/jedi/issues/713

**Comment by Carreau** [comment]
Travis tests are failing, we should add jedi either as a required (or optional) dependency in `setup.py`. 

@takluyver do we make it optional or required ? If we make it required we can simplify the code in `completer.py` and actually fail on import error.

File: IPython/core/completer.py [hunk 4]
```diff
@@ -765,9 +775,59 @@ def magic_matches(self, text):
...
+                no__name = lambda txt: not txt.startswith('__')
+            else:
+                # true if txt is _not_ a _ name, false otherwise:
+                no__name = lambda txt: not txt.startswith('_')
+            completion_text = filter(no__name, completion_text)
+
+        matches = [before + c_text for c_text in completion_text]
+        return matches
+
     def python_matches(self,text):
```
Comment (liukelly) [inline IPython/core/completer.py]: To be deprecated.

File: IPython/core/completer.py (hunk 3 above)
Comment (liukelly) [inline IPython/core/completer.py]: It was from http://jedi.jedidjah.ch/en/latest/_modules/jedi/utils.html. It seems to work without it, so maybe we can take it out, but I don't know why it's there.

Comment (takluyver) [inline IPython/core/completer.py]: I'm not sure why that Jedi code does it, but IPython already adds the CWD to `sys.path`, so I'm pretty sure we don't need to do it here.

**Comment by takluyver** [comment]
Once we're satisfied that jedi produces good results, and we're willing to get rid of our own code, I'd say we should make it a regular dependency, not an optional thing.

We should make sure that it doesn't mess up completions with things like magics. We have our own machinery to do completions in such cases, but we should ensure that jedi doesn't do something bizarre with our extra syntax.

File: IPython/terminal/ptshell.py [hunk 5]
```diff
@@ -49,8 +49,7 @@ def get_completions(self, document, complete_event):
         )
         start_pos = -len(used)
         for m in matches:
-            yield Completion(m, start_position=start_pos)
-
+            yield Completion(m, start_position=start_pos, display_meta=m[::-1])
```
Comment (liukelly) [inline IPython/terminal/ptshell.py]: completion.py in Python Prompt Toolkit [[link](https://github.com/jonathanslenders/python-prompt-toolkit/blob/master/prompt_toolkit/completion.py)]

Comment (Carreau) [inline IPython/terminal/ptshell.py]: Sweet ! Just for other information, the second column is the completion text reversed for now, as Jedi return a wrong value for now. We planned on adding extra info in this field later.

**Comment by Carreau** [comment]
I've made some improvements for the greedy completion, so that things like `a[0].<tab>` works: https://github.com/liukelly/ipython/pull/1,

This will also need a rebase. 

@liukelly tell us if you are not comfortable with  rebase. 

until jedi is fixed, I think we should just not use the `, display_meta=m[::-1]` part, and open an issue to remind us to check that later.

**Comment by liukelly** [comment]
Sorry for the delay, but I think I resolved all the conflicts.

**Comment by Carreau** [comment]
> Sorry for the delay, but I think I resolved all the conflicts.

No worries. Will have a look.

**Comment by Carreau** [comment]
Looks good to me. is the addition of `path.py` in `setup.py` for tests necessary ?

**Comment by takluyver** [comment]
> is the addition of path.py in setup.py for tests necessary ? 

Probably a mix up in resolving some merge conflicts - that was added in master to make tests pass, and then the need for it was eliminated, so the dep was removed again.

**Comment by Carreau** [comment]
> Probably a mix up in resolving some merge conflicts - that was added in master to make tests pass, and then the need for it was eliminated, so the dep was removed again.

Ok. 

I'll you review that before going to sleep . I know the sun never set on the British empire, but it's not a reason for you not to get some rest.

**Comment by liukelly** [comment]
> is the addition of path.py in setup.py for tests necessary ?

Sorry, I think that came from liukelly/ipython@2f228c0d920cc88cb52b77ba104f1f9c8594f2f1. I don't think we need it.

**Comment by Carreau** [comment]
> Sorry, I think that came from liukelly/ipython@2f228c0. I don't think we need it.
>  Show all checks

No problem, 
let's merge as it, get it in the wild to test. And I'll remove path.py in a subsequent PR. 

Thanks !

**Pull request merged**